### PR TITLE
fix tikv#218

### PR DESCRIPTION
**Pull request by morefreeze**
See pingcap/tikv#218

Extra change:
1. Scan error handler
2. log more detail about error
3. remove TiKVIter code

### CODE REVIEW

**Comment by morefreeze** [comment]
@disksing PTAL

File: txn.go [hunk 1]
```diff
@@ -98,28 +98,31 @@ func (txn *tikvTxn) doCommit() error {
 	// so don't need pick primary key.
 	cmdPrewriteReq := new(pb.CmdPrewriteRequest)
 	// commitKeys for commit request keys
-	var commitKeys [][]byte
+	var commitKeys []*pb.KeyAddress
 	err := txn.us.WalkBuffer(func(k kv.Key, v []byte) error {
```
Comment (disksing) [inline txn.go]: unchecked error?

File: txn.go [hunk 2]
```diff
@@ -98,28 +98,31 @@ func (txn *tikvTxn) doCommit() error {
...
 		return nil
 	})
-	cmdPrewriteReq.Locks = txn.lockKeys
-	commitKeys = append(commitKeys, txn.lockKeys...)
+	for _, lockKey := range txn.lockKeys {
+		row := &pb.KeyAddress{Key: lockKey}
+		cmdPrewriteReq.Locks = append(cmdPrewriteReq.Locks, row)
+		commitKeys = append(commitKeys, row)
+	}
 	// If txn keys array is empty then no need send rpc.
```
Comment (disksing) [inline txn.go]: If the transaction has only lock keys (no put/delete), we do not need send rpc, either.

File: txn.go (hunk 1 above)
Comment (morefreeze) [inline txn.go]: Nice catch!

File: txn.go (hunk 2 above)
Comment (morefreeze) [inline txn.go]: Yeah, that make sense.

**Comment by morefreeze** [comment]
@disksing PTAL

**Comment by disksing** [comment]
LGTM

**Comment by morefreeze** [comment]
@siddontang @tangliu PTAL

File: client.go [hunk 3]
```diff
@@ -105,7 +105,7 @@ func (c *Client) Close() error {
 
 func (c *Client) doSend(conn net.Conn, req *pb.Request) (*pb.Response, error) {
 	curMsgID := atomic.AddUint64(&c.msgID, 1)
-	log.Debugf("Send request msgID[%d] type[%s] req[%q]", curMsgID, req.GetType(), req)
+	log.Debugf("Send request msgID[%d] type[%s] req[%#v]", curMsgID, req.GetType(), req)
```
Comment (siddontang) [inline client.go]: mabye %s is better than %v for printing protobuf message. you can try and compare the output.

Comment (morefreeze) [inline client.go]: Yes, %s is better.

**Comment by morefreeze** [comment]
@siddontang PTAL I review all `%#v` and replace right argument.

**Comment by siddontang** [comment]
LGTM

**Pull request merged**